File: assets/scripts/maze/mazeGenerator.ts
import { randomRangeInt } from 'cc';

export namespace Maze_MazeGenerator
{
    enum eNeighbourType
    {
        TOP = 0,
        RIGHT = 1,
        BOTTOM = 2,
        LEFT = 3
    }

    export class MazeNode
    {
        private _x:number;
        public get x():number
        {
            return this._x;
        }

        private _y:number;
        public get y():number
        {
            return this._y;
        }

        constructor(x:number, y:number)
        {
            this._x = x;
            this._y = y;
        }

        private _topSibling:MazeNode|null = null;
        public set topSibling(val:MazeNode|null)
        {
            this._topSibling = val;
        }
        public get topSibling():MazeNode|null
        {
            return this._topSibling;
        }

        private _rightSibling:MazeNode|null = null;
        public set rightSibling(val:MazeNode|null)
        {
            this._rightSibling = val;
        }
        public get rightSibling():MazeNode|null
        {
            return this._rightSibling;
        }
        
        private _bottomSibling:MazeNode|null = null;
        public set bottomSibling(val:MazeNode|null)
        {
            this._bottomSibling = val;
        }
        public get bottomSibling():MazeNode|null
        {
            return this._bottomSibling;
        }

        private _leftSibling:MazeNode|null = null;
        public set leftSibling(val:MazeNode|null)
        {
            this._leftSibling = val;
        }
        public get leftSibling():MazeNode|null
        {
            return this._leftSibling;
        }

        private _visited:boolean = false;
        public set visited(val:boolean)
        {
            this._visited = val;
        }
        public get visited():boolean
        {
            return this._visited;
        }

        private _topWall:boolean = true;
        public set topWall(val:boolean)
        {
            this._topWall = val;
        }
        public get topWall():boolean
        {
            return this._topWall;
        }

        private _rightWall:boolean = true;
        public set rightWall(val:boolean)
        {
            this._rightWall = val;
        }
        public get rightWall():boolean
        {
            return this._rightWall;
        }

        private _bottomWall:boolean = true;
        public set bottomWall(val:boolean)
        {
            this._bottomWall = val;
        }
        public get bottomWall():boolean
        {
            return this._bottomWall;
        }

        private _leftWall:boolean = true;
        public set leftWall(val:boolean)
        {
            this._leftWall = val;
        }
        public get leftWall():boolean
        {
            return this._leftWall;
        }

        public getAllNotVisitedNeighbours():Map<eNeighbourType,MazeNode>
        {
            var neighbours:Map<eNeighbourType,MazeNode> = new Map<eNeighbourType,MazeNode>();

            if(null != this.leftSibling && false == this.leftSibling.visited)
            {
                neighbours.set(eNeighbourType.LEFT, this.leftSibling);
            }

            if(null != this.rightSibling && false == this.rightSibling.visited)
            {
                neighbours.set(eNeighbourType.RIGHT, this.rightSibling);
            }

            if(null != this.topSibling && false == this.topSibling.visited)
            {
                neighbours.set(eNeighbourType.TOP, this.topSibling);
            }

            if(null != this.bottomSibling && false == this.bottomSibling.visited)
            {
                neighbours.set(eNeighbourType.BOTTOM, this.bottomSibling);
            }

            return neighbours;
        }

        public getAllVisitedNeighbours():Map<eNeighbourType,MazeNode>
        {
            var neighbours:Map<eNeighbourType,MazeNode> = new Map<eNeighbourType,MazeNode>();

            if(null != this.leftSibling && true == this.leftSibling.visited)
            {
                neighbours.set(eNeighbourType.LEFT, this.leftSibling);
            }

            if(null != this.rightSibling && true == this.rightSibling.visited)
            {
                neighbours.set(eNeighbourType.RIGHT, this.rightSibling);
            }

            if(null != this.topSibling && true == this.topSibling.visited)
            {
                neighbours.set(eNeighbourType.TOP, this.topSibling);
            }

            if(null != this.bottomSibling && true == this.bottomSibling.visited)
            {
                neighbours.set(eNeighbourType.BOTTOM, this.bottomSibling);
            }

            return neighbours;
        }

        public getRandomNotVisitedNeighbour():[eNeighbourType,MazeNode]|undefined
        {
            var neighbours = this.getAllNotVisitedNeighbours();

            if(neighbours.size > 0)
            {
                var targetIndex = randomRangeInt(0, neighbours.size);
                var i:number = 0;
                for(var neighbourInfo of neighbours)
                {
                    if(i == targetIndex)
                    {
                        return neighbourInfo;
                        break;
                    }
                    
                    i += 1;
                }
            }
            else
            {
                return undefined;
            }
        }

        public getRandomVisitedNeighbour():[eNeighbourType,MazeNode]|undefined
        {
            var neighbours = this.getAllVisitedNeighbours();

            if(neighbours.size > 0)
            {
                var targetIndex = randomRangeInt(0, neighbours.size);
                var i:number = 0;
                for(var neighbourInfo of neighbours)
                {
                    if(i == targetIndex)
                    {
                        return neighbourInfo;
                        break;
                    }
                    
                    i += 1;
                }
            }
            else
            {
                return undefined;
            }
        }
    }

    export function generateMaze(width:number, height:number):MazeNode[][]
    {
        var result:MazeNode[][] = [];

        if(width > 0 && height > 0)
        {
            // create all nodes
            for(var row:number = 0; row < height; ++row)
            {
                result.push([]);

                for(var column:number = 0; column < width; ++column)
                {
                    var mazeNode = new MazeNode(column, row);                
                    result[row].push(mazeNode);
                }
            }

            // assign nodes to each other
            for(var row:number = 0; row < height; ++row)
            {
                for(var column:number = 0; column < width; ++column)
                {
                    if(row > 0)
                    {
                        result[row][column].topSibling = result[row-1][column];
                    }

                    if(column > 0)
                    {
                        result[row][column].leftSibling = result[row][column-1];
                    }

                    if(row < (height - 1))
                    {
                        result[row][column].bottomSibling = result[row+1][column];
                    }

                    if(column != (width - 1))
                    {
                        result[row][column].rightSibling = result[row][column+1];
                    }
                }
            }

            var startNode:MazeNode = result[randomRangeInt(0,height)][randomRangeInt(0,width)];
            var frontier:Set<MazeNode> = new Set<MazeNode>();

            var neighbours = startNode.getAllNotVisitedNeighbours();

            for(var neighbourInfo of neighbours)
            {
                frontier.add(neighbourInfo[1]);
            }
            startNode.visited = true;

            while(0 != frontier.size)
            {
                // select random frontier
                var randomFrontierIdx = randomRangeInt(0,frontier.size);
                var randomFrontier:MazeNode = result[0][0];
                
                var i:number = 0;
                for(var frontierElemnt of frontier)
                {
                    if(i == randomFrontierIdx)
                    {
                        randomFrontier = frontierElemnt;
                        break;
                    }
                    
                    i += 1;
                }

                // select "break the wall visited neighbour
                var randomVisitedNeighbourInfo = randomFrontier.getRandomVisitedNeighbour();

                if(null != randomVisitedNeighbourInfo)
                {
                    switch(randomVisitedNeighbourInfo[0])
                    {
                        case eNeighbourType.BOTTOM:
                            randomFrontier.bottomWall = false;
                            randomVisitedNeighbourInfo[1].topWall = false;
                            break;
                        case eNeighbourType.RIGHT:
                            randomFrontier.rightWall = false;
                            randomVisitedNeighbourInfo[1].leftWall = false;
                            break;
                        case eNeighbourType.TOP:
                            randomFrontier.topWall = false;
                            randomVisitedNeighbourInfo[1].bottomWall = false;
                            break;
                        case eNeighbourType.LEFT:
                            randomFrontier.leftWall = false;
                            randomVisitedNeighbourInfo[1].rightWall = false;
                            break;
                    }

                    randomFrontier.visited = true;
                    
                    var neighbours = randomFrontier.getAllNotVisitedNeighbours();

                    for(var neighbourInfo of neighbours)
                    {
                        frontier.add(neighbourInfo[1]);
                    }

                    frontier.delete(randomFrontier);
                }
                else
                {
                    throw("Did not find visited neighbours!");
                }
            }
        }

        return result;
    }
}

File: Maze_Unity/Assets/Scripts/Maze/MazeGenerator.cs
using System.Collections.Generic;
using System;
using Maze_Common;

namespace Maze_MazeGenerator
{
    enum eNeighbourType
    {
        TOP = 0,
        RIGHT = 1,
        BOTTOM = 2,
        LEFT = 3
    }

    class MazeNode
    {
        private int mX;
        public int X { get => mX; }

        private int mY;
        public int Y { get => mY; }

        public MazeNode(int x, int y)
        {
            this.mX = x;
            this.mY = y;
        }

        private MazeNode mTopSibling = null;
        public MazeNode TopSibling { get => mTopSibling; set => mTopSibling = value; }

        private MazeNode mRightSibling = null;
        public MazeNode RightSibling { get => mRightSibling; set => mRightSibling = value; }

        private MazeNode mBottomSibling = null;
        public MazeNode BottomSibling { get => mBottomSibling; set => mBottomSibling = value; }

        private MazeNode mLeftSibling = null;
        public MazeNode LeftSibling { get => mLeftSibling; set => mLeftSibling = value; }

        private bool mVisited = false;
        public bool Visited { get => mVisited; set => mVisited = value; }

        private bool mTopWall = true;
        public bool TopWall { get => mTopWall; set => mTopWall = value; }

        private bool mRightWall = true;
        public bool RightWall { get => mRightWall; set => mRightWall = value; }

        private bool mBottomWall = true;
        public bool BottomWall { get => mBottomWall; set => mBottomWall = value; }

        private bool mLeftWall = true;
        public bool LeftWall { get => mLeftWall; set => mLeftWall = value; }

        public Dictionary<eNeighbourType, MazeNode> getAllNotVisitedNeighbours()
        {
            Dictionary<eNeighbourType, MazeNode> neighbours = new Dictionary<eNeighbourType, MazeNode>();

            if (null != this.mLeftSibling && false == this.mLeftSibling.mVisited)
            {
                neighbours.Add(eNeighbourType.LEFT, this.mLeftSibling);
            }

            if (null != this.mRightSibling && false == this.mRightSibling.mVisited)
            {
                neighbours.Add(eNeighbourType.RIGHT, this.mRightSibling);
            }

            if (null != this.mTopSibling && false == this.mTopSibling.mVisited)
            {
                neighbours.Add(eNeighbourType.TOP, this.mTopSibling);
            }

            if (null != this.mBottomSibling && false == this.mBottomSibling.mVisited)
            {
                neighbours.Add(eNeighbourType.BOTTOM, this.mBottomSibling);
            }

            return neighbours;
        }

        public Dictionary<eNeighbourType, MazeNode> getAllVisitedNeighbours()
        {
            Dictionary<eNeighbourType, MazeNode> neighbours = new Dictionary<eNeighbourType, MazeNode>();

            if (null != this.mLeftSibling && true == this.mLeftSibling.mVisited)
            {
                neighbours.Add(eNeighbourType.LEFT, this.mLeftSibling);
            }

            if (null != this.mRightSibling && true == this.mRightSibling.mVisited)
            {
                neighbours.Add(eNeighbourType.RIGHT, this.mRightSibling);
            }

            if (null != this.mTopSibling && true == this.mTopSibling.mVisited)
            {
                neighbours.Add(eNeighbourType.TOP, this.mTopSibling);
            }

            if (null != this.mBottomSibling && true == this.mBottomSibling.mVisited)
            {
                neighbours.Add(eNeighbourType.BOTTOM, this.mBottomSibling);
            }

            return neighbours;
        }

        public Tuple<eNeighbourType, MazeNode> getRandomNotVisitedNeighbour()
        {
            var neighbours = this.getAllNotVisitedNeighbours();

            if (neighbours.Count > 0)
            {
                int targetIndex = Common.randomRangeInt(0, neighbours.Count);
                int i = 0;
                foreach (var neighbourInfo in neighbours)
                {
                    if (i == targetIndex)
                    {
                        return Tuple.Create(neighbourInfo.Key, neighbourInfo.Value);
                    }

                    i += 1;
                }
            }
            else
            {
                return null;
            }

            return null;
        }

        public Tuple<eNeighbourType, MazeNode> getRandomVisitedNeighbour()
        {
            var neighbours = this.getAllVisitedNeighbours();

            if (neighbours.Count > 0)
            {
                var targetIndex = Common.randomRangeInt(0, neighbours.Count);
                int i = 0;
                foreach (var neighbourInfo in neighbours)
                {
                    if (i == targetIndex)
                    {
                        return Tuple.Create(neighbourInfo.Key, neighbourInfo.Value);
                    }

                    i += 1;
                }
            }
            else
            {
                return null;
            }

            return null;
        }
    }

    class MazeGenerator
    {
        List<List<MazeNode>> generateMaze(int width, int height)
        {
            List<List<MazeNode>> result = new List<List<MazeNode>>();

            if (width > 0 && height > 0)
            {
                // create all nodes
                for (int row = 0; row < height; ++row)
                {
                    result.Add(new List<MazeNode>());

                    for (int column = 0; column < width; ++column)
                    {
                        var mazeNode = new MazeNode(column, row);
                        result[row].Add(mazeNode);
                    }
                }
            }

            // assign nodes to each other
            for (int row = 0; row < height; ++row)
            {
                for (int column = 0; column < width; ++column)
                {
                    if (row > 0)
                    {
                        result[row][column].TopSibling = result[row - 1][column];
                    }

                    if (column > 0)
                    {
                        result[row][column].LeftSibling = result[row][column - 1];
                    }

                    if (row < (height - 1))
                    {
                        result[row][column].BottomSibling = result[row + 1][column];
                    }

                    if (column != (width - 1))
                    {
                        result[row][column].RightSibling = result[row][column + 1];
                    }
                }
            }

            MazeNode startNode = result[Common.randomRangeInt(0, height)][Common.randomRangeInt(0, width)];
            HashSet<MazeNode> frontier = new HashSet<MazeNode>();

            var neighbours = startNode.getAllNotVisitedNeighbours();

            foreach (var neighbourInfo in neighbours)
            {
                frontier.Add(neighbourInfo.Value);
            }
            startNode.Visited = true;

            while (0 != frontier.Count)
            {
                // select random frontier
                var randomFrontierIdx = Common.randomRangeInt(0, frontier.Count);
                MazeNode randomFrontier = result[0][0];

                int i = 0;
                foreach (var frontierElemnt in frontier)
                {
                    if (i == randomFrontierIdx)
                    {
                        randomFrontier = frontierElemnt;
                        break;
                    }

                    i += 1;
                }

                // select "break the wall" visited neighbour
                var randomVisitedNeighbourInfo = randomFrontier.getRandomVisitedNeighbour();

                if (null != randomVisitedNeighbourInfo)
                {
                    switch (randomVisitedNeighbourInfo.Item1)
                    {
                        case eNeighbourType.BOTTOM:
                            randomFrontier.BottomWall = false;
                            randomVisitedNeighbourInfo.Item2.TopWall = false;
                            break;
                        case eNeighbourType.RIGHT:
                            randomFrontier.RightWall = false;
                            randomVisitedNeighbourInfo.Item2.LeftWall = false;
                            break;
                        case eNeighbourType.TOP:
                            randomFrontier.TopWall = false;
                            randomVisitedNeighbourInfo.Item2.BottomWall = false;
                            break;
                        case eNeighbourType.LEFT:
                            randomFrontier.LeftWall = false;
                            randomVisitedNeighbourInfo.Item2.RightWall = false;
                            break;
                    }

                    randomFrontier.Visited = true;

                    var neighbours_ = randomFrontier.getAllNotVisitedNeighbours();

                    foreach (var neighbourInfo in neighbours_)
                    {
                        frontier.Add(neighbourInfo.Value);
                    }

                    frontier.Remove(randomFrontier);
                }
                else
                {
                    throw new System.Exception("Did not find visited neighbours!");
                }
            }

            return result;
        }
    }
}